CI Infra Docs (#7985)
dask/dask: docs/source/develop.rst

``git commit --no-verify`` or with the short version ``git commit -n``.

> Dask CI Infrastructure

Reviewer: Can we move this section to the end of this page? Currently it's between the "Contributing to Code" and "Contributing to Documentation" sections, which seems a little out of place

----------------------

Github Actions
~~~~~~~~~~~~~~

Dask uses Github Actions for Continuous Integration (CI) testing for each PR.
The actions will run the test suite across a variety of Python versions, Operating
Systems, and NumPy and Pandas versions.  Additionally, Dask can upstream Xarray tests

> with the commit message ``test-upstream``

Author: +1

The CI workflows for Github Actions are defined in
`.github/workflows <https://github.com/dask/dask/tree/main/.github/workflows>`_
with additonal scripts and metadata located in `continuous_integration
<https://github.com/dask/dask/tree/main/continuous_integration`_

GPU CI
~~~~~~

As Dask seamlessly work with GPUs, Dask will also test PRs with a GPU enabled
CI environment provided by NVIDIA: `gpuCI <https://gpuci.gpuopenanalytics.com/>`_ .
Unlike Github Actions, the CI environment for gpuCI is controlled with a `docker image
<https://github.com/rapidsai/dask-build-environment/>`_ .  When making commits

> to the repo, a new image is built.  The docker image building process can be

Author: Updated in ba5527359

monitored
`here
<https://gpuci.gpuopenanalytics.com/job/dask/job/dask-build-environment/job/branch/job/dask-build-env-main/>`_
.  Note, the ``dask-build-environment`` has two separate Dockerfiles for Dask
and Distributed similiarlly, gpuCI will run for both `Dask
https://gpuci.gpuopenanalytics.com/job/dask/job/dask/job/prb/job/dask-prb/>`_
and `Distributed
<https://gpuci.gpuopenanalytics.com/job/dask/job/distributed/job/prb/job/distributed-prb/>`_

For each PR, gpuCI will run all tests decorated with the pytest marker
`@pytest.mark.gpu`.  This is configured in the `gpuci folder
<https://github.com/dask/dask/tree/main/continuous_integration/gpuci>`_ .

> Like Github Actions, gpuCI will not run when first time authors to Dask or

Reviewer: ```suggestion
Like Github Actions, gpuCI will not run when first time contributors to Dask or
```

> Distributed submit PRs.  In this case, a gpuCI bot will comment on the PR:

Reviewer: ```suggestion
Distributed submit PRs.  In this case, the gpuCI bot will comment on the PR:
```

> Can one of the admins verify this patch?

> Dask Maintainers can approved build these PRs with following choices:

Reviewer: ```suggestion
Dask Maintainers can then approve gpuCI builds for these PRs with following choices:
```

> - To approve PR author for just the current PR, comment the command:
>   - ``ok to test``
> - To approve PR author for all PRs in the future, comment the command:
>   - ``add to allowlist``

Reviewer: This seems slightly more clear to me, though feel free to push back

```suggestion
- To only approve the PR contributor for the current PR, leave a comment which states ``ok to test``
- To approve the current PR and all future PRs from the contributor, leave a comment which states ``add to allowlist``
```

`rapidsai/dask-build-environment <https://github.com/rapidsai/dask-build-environment/>`_
docker image.  When making commits to the
`dask-build-environment repo <https://github.com/rapidsai/dask-build-environment/>`_ , a new image is built.
The docker image building process can be monitored
`here <https://gpuci.gpuopenanalytics.com/job/dask/job/dask-build-environment/job/branch/job/dask-build-env-main/>`_.
Note, the ``dask-build-environment`` has two separate Dockerfiles for Dask
and Distributed similiarlly, gpuCI will run for both `Dask
<https://gpuci.gpuopenanalytics.com/job/dask/job/dask/job/prb/job/dask-prb/>`_
and `Distributed
<https://gpuci.gpuopenanalytics.com/job/dask/job/distributed/job/prb/job/distributed-prb/>`_

For each PR, gpuCI will run all tests decorated with the pytest marker
``@pytest.mark.gpu``.  This is configured in the `gpuci folder
<https://github.com/dask/dask/tree/main/continuous_integration/gpuci>`_ .
Like Github Actions, gpuCI will not run when first time contributors to Dask or
Distributed submit PRs.  In this case, the gpuCI bot will comment on the PR:

.. note:: Can one of the admins verify this patch?

.. image:: images/gputester-msg.png

> Dask Maintainers can then approve gpuCI builds for these PRs with following choices:

Reviewer 2: What kind of github permission level does this require? Afaik, there are `Read`, `Triage`, `Write`, `Maintain` and `Admin` privileges, we might not be using all of them. From this passage, I would guess it is `Maintain`? see https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/repository-permission-levels-for-an-organization

Related question: What is the default permission for dask org `member`s?

Author: Currently, this is limited to folks who [Dask owners](https://github.com/orgs/dask/people?query=role%3Aowner) but we don't have to be limited to that group.  gpuCI can be manually manipulated to include additional folks who may not be a Dask owner but still need/require/want this capability.  

I think default permissions for dask members is the same as public permissions: read/create PSs/issues/comments/etc but no write access
